### PR TITLE
Remove support for 'ref scoped'

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/62338

Relates to test plan https://github.com/dotnet/roslyn/issues/59194

### CODE REVIEW

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review please.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs [hunk 1]
```diff
@@ -541,14 +540,28 @@ internal static void CheckParameterModifiers(BaseParameterSyntax parameter, Bind
                     case SyntaxKind.ScopedKeyword when !parsingFunctionPointerParams:
                         ModifierUtils.CheckScopedModifierAvailability(parameter, modifier, diagnostics);
                         {
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:542]: >{ [](http://example.com/codeflow?start=24&length=1)

Minor: This enclosing block is no longer necessary.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs [hunk 2]
```diff
@@ -10088,8 +10088,6 @@ class Program
 {
     static R F1(ref R r) { Console.WriteLine(r._i); return new R(); }
     static R F2(scoped ref R r) { Console.WriteLine(r._i); return new R(); }
-    static R F3(ref scoped R r) { Console.WriteLine(r._i); return new R(); }
-    static R F4(scoped ref R r) { Console.WriteLine(r._i); return new R(); }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:10092]: >static R F4(scoped ref R r) { Console.WriteLine(r._i); return new R(); } [](http://example.com/codeflow?start=4&length=72)

Consider keeping this case and `d4` below, to verify the synthesized delegate is shared between `d2` and `d4`.

**Comment by cston** [comment]
>     public static implicit operator B<T>(in scoped A<T> a) => default;

Consider keeping this and changing the parameter to `scoped in A<T> a`, to ensure we're testing operators.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:6692 in 95c778a. [](commit_id = 95c778a97ecd0c872cced209798f2f24fdee5b6a, deletion_comment = True)

**Comment by cston** [comment]
>     public void M4(scoped ref R<T> r) { }

Let's test `ref R<T> r` and `scoped ref R<T> r` pairs instead. In short, keep the `M3()` methods but change `ref scoped` to simply `ref`.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8741 in 95c778a. [](commit_id = 95c778a97ecd0c872cced209798f2f24fdee5b6a, deletion_comment = True)

**Comment by cston** [comment]
>     public abstract R<T> F3(ref scoped R<T> r);

Let's test `ref` rather than `ref scoped`.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8800 in 95c778a. [](commit_id = 95c778a97ecd0c872cced209798f2f24fdee5b6a, deletion_comment = True)

**Comment by cston** [comment]
>     R<T> F3(ref scoped R<T> r);

Let's test `ref` rather than `ref scoped`.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8860 in 95c778a. [](commit_id = 95c778a97ecd0c872cced209798f2f24fdee5b6a, deletion_comment = True)

**Comment by cston** [comment]
>         var f = (ref scoped Unknown u) => { };

Change to `scoped ref`?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8968 in 95c778a. [](commit_id = 95c778a97ecd0c872cced209798f2f24fdee5b6a, deletion_comment = False)

**Comment by cston** [comment]
> 

Extra blank line.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12301 in 95c778a. [](commit_id = 95c778a97ecd0c872cced209798f2f24fdee5b6a, deletion_comment = False)

**Comment by cston** [comment]
>             var comp = CreateCompilationWithSpanAndMemoryExtensions(source);

`CreateCompilation` here and for other calls in this test.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12310 in 95c778a. [](commit_id = 95c778a97ecd0c872cced209798f2f24fdee5b6a, deletion_comment = False)

File: src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs [hunk 3]
```diff
@@ -8338,8 +8322,7 @@ class Program
 {
     static void M(R r0)
     {
-        scoped R r1;
-        ref readonly scoped R r2 = ref r1;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:8342]: >ref readonly scoped [](http://example.com/codeflow?start=8&length=19)

We should include a `ref readonly` example. Perhaps change `r3` to `scoped ref readonly R r3;`

File: src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb [hunk 4]
```diff
@@ -5383,8 +5376,7 @@ class Program
 {
     static void M(R r0)
     {
-        scoped R r1;
-        ref readonly scoped R r2 = ref r1;
+        scoped R r1 = r0;
         scoped ref R r3 = ref r0;
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb:5380]: >scoped ref R r3 [](http://example.com/codeflow?start=8&length=15)

`scoped ref readonly R r3`

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

**Comment by cston** [comment]
>     public void M3(ref scoped R<string> r) { } // 3

`scoped ref`?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8292 in 2d8194d. [](commit_id = 2d8194dbb2907c4cc3886612ae50798c7eafceeb, deletion_comment = False)

**Comment by cston** [comment]
>     public override object this[scoped R<string> r] { get { return null; } set { } } // 5

`F3(scoped ref R<string> r)` and `F4(scoped ref R<string> r)` instead?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8353 in 2d8194d. [](commit_id = 2d8194dbb2907c4cc3886612ae50798c7eafceeb, deletion_comment = False)

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 5]
```diff
@@ -7343,9 +7076,33 @@ public void DelegateReturnTypeScope(LanguageVersion langVersion)
 ";
             var comp = CreateCompilation(source, parseOptions: TestOptions.Regular.WithLanguageVersion(langVersion));
             comp.VerifyEmitDiagnostics(
-                // (2,14): error CS0106: The modifier 'scoped' is not valid for this item
+                // (2,14): error CS0246: The type or namespace name 'scoped' could not be found (are you missing a using directive or an assembly reference?)
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:7079]: This is an extremely bad parse. I think we should be able to do better.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 6]
```diff
@@ -9448,19 +8895,63 @@ class Program
...
+    static void F(S s)
+    {{
+        {refModifier} scoped S s1 = ref s;
         scoped {refModifier} scoped S s3 = ref s;
     }}
 }}";
             var comp = CreateCompilation(source);
             comp.VerifyDiagnostics(
-                // (6,29): error CS9048: The 'scoped' modifier can be used for refs and ref struct values only.
+                // (6,22): error CS0246: The type or namespace name 'scoped' could not be found (are you missing a using directive or an assembly reference?)
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:8922]: Same here. We forbid types named `scoped` in C# 11, right? This should be unambiguously a bad modifier, not a type name.

**Review by 333fred — COMMENTED** [review]
Done review pass. I went through some of the tests, and I don't think the approach we're taking towards misplaced `scoped` is good. We should be able to recognize that it's a misplaced modifier and give a proper error, rather than dumping the user into a bad state. In particular, I expect people accidentally typing `ref scoped` instead of `scoped ref` to a very common failure (I still don't remember off the top of my head if it's `ref readonly` or `readonly ref`), so we need to do better for the error recovery here.

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md [hunk 7]
```diff
@@ -1,5 +1,22 @@
 # This document lists known breaking changes in Roslyn after .NET 6 all the way to .NET 7.
 
+## `scoped` treated as a contextual keyword
+
+***Introduced in Visual Studio 2022 version 17.4.*** Starting in C# 11, types cannot be named `scoped`. The compiler will report an error on all such type names. To work around this, the type name and all usages must be escaped with an `@`:
```
Comment (jcouv) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:5]: :memo: the full implementation of this break will come shortly

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md [hunk 8]
```diff
@@ -1,5 +1,22 @@
 # This document lists known breaking changes in Roslyn after .NET 6 all the way to .NET 7.
 
+## `scoped` treated as a contextual keyword
```
Comment (cston) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:3]: Should this be "Types cannot be named `scoped`" instead, assuming that captures the break, for consistency with the changes for `file` and `required`?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 9]
```diff
@@ -7415,35 +7391,61 @@ public void LocalScope_04()
...
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "=").WithArguments("=").WithLocation(2, 15),
                 // (4,1): error CS8936: Feature 'ref fields' is not available in C# 10.0. Please use language version 11.0 or greater.
-                // scoped ref scoped s4 = ref s1;
+                // scoped scoped s4 = default; // 2
                 Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion10, "scoped").WithArguments("ref fields", "11.0").WithLocation(4, 1),
-                // (5,12): warning CS8981: The type name 'scoped' only contains lower-cased ascii characters. Such names may become reserved for the language.
-                // ref struct scoped { }
-                Diagnostic(ErrorCode.WRN_LowerCaseTypeName, "scoped").WithArguments("scoped").WithLocation(5, 12));
+                // (4,15): warning CS0219: The variable 's4' is assigned but its value is never used
+                // scoped scoped s4 = default; // 2
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:7408]: Are we expecting an error on types named `scoped` in a future PR? Presumably what's happening is that the parser isn't treating the second `scoped` as a contextual keyword here?

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs [hunk 10]
```diff
@@ -1546,23 +1847,192 @@ public void Local_05(LanguageVersion langVersion)
...
+            EOF();
+        }
+
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(LanguageVersion.CSharp11)]
+        public void Local_06(LanguageVersion langVersion)
+        {
+            string source =
+@"scoped.nested a;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs:1929]: >scoped.nested [](http://example.com/codeflow?start=2&length=13)

It seems odd that `scoped.nested` is treated as a type name in `scoped.nested a;` but not in `ref scoped.nested b;`.

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs [hunk 11]
```diff
@@ -2753,8 +3295,19 @@ public void Using_01(LanguageVersion langVersion)
             string source =
 @"using scoped s;
 using ref scoped r;
+using ref scoped r r2;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs:3298]: >using ref scoped r r2 [](http://example.com/codeflow?start=0&length=21)

Covered in `Using_02_RefScoped`?

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs [hunk 12]
```diff
@@ -2874,9 +3469,29 @@ public void Using_03(LanguageVersion langVersion)
         {
             string source =
 @"await using scoped s;
+await using ref scoped;
 await using ref scoped r;
+await using ref scoped r r2;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs:3474]: Covered in `Using_04_RefScoped`?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:7408]: I don't expect a blanket error on referencing types names `scoped`. I only expect that anytime we're looking for `scoped`, we'll prefer the keyword interpretation to the identifier interpretation.
The behavior for `scoped scoped s4` is an existing issue. Filed tracking issue: [https://github.com/dotnet/roslyn/issues/62950](https://github.com/dotnet/roslyn/issues/62950)

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationScopeParsingTests.cs:1929]: Filed follow-up issue: [https://github.com/dotnet/roslyn/issues/62950](https://github.com/dotnet/roslyn/issues/62950)

**Review by cston — APPROVED** [review]

**Review by 333fred — APPROVED** [review]

**Pull request merged**